hil/ext/switches/ovs.py

```diff
@@ -0,0 +1,266 @@
+# License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS
+# IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied.  See the License for the specific language
+# governing permissions and limitations under the License.
+"""A switch driver for the juniper QFX series.
+
+But in theory can work for any Juniper switch that supports NETCONF.
+Driver uses the python library PyEZ made available by Juniper.
+By default it sets NETCONF session over ssh, i.e. make RPC calls over ssh.
+For this to work, make sure that the Juniper switch is configured
+appropriately to recieve NETCONF connections. Its default port is 830.
+"""
+
+import re
+import logging
+import schema
+import os
+import ast
+from commands import getstatusoutput
+
+from hil.model import db, Switch, BigIntegerType
+from hil.migrations import paths
+from os.path import dirname, join
+logger = logging.getLogger(__name__)
+paths[__name__] = join(dirname(__file__), 'migrations', 'ovs')
+
+
+class VlanError(Exception):
+    """ Raise this exception when vlan cannot be added to the port."""
+
+
+class NotImplementedYet(Exception):
+    """ Raise this exception for functions that require
+    more information to implement.
+    """
+
+
+class PortInfoNotFound(Exception):
+    """ Raise this exception when this driver is unable to fetch
+    information for a given port.
+    """
+
+
+class OVS_RequestFailed(Exception):
+    """ Raise this exception when a requestOVS failure
+    cannot does not follow into any of the above exceptions.
+    """
+
+
+class Ovs(Switch):
+    """ Driver for openvswitch. """
+    api_name = 'http://schema.massopencloud.org/haas/v0/switches/ovs'
+    dir_name = os.path.dirname(__file__)
+    __mapper_args__ = {
+        'polymorphic_identity': api_name,
+    }
+
+    id = db.Column(
+            BigIntegerType, db.ForeignKey('switch.id'), primary_key=True
+            )
+    hostname = db.Column(db.String, nullable=False)
+    username = db.Column(db.String, nullable=False)
+    password = db.Column(db.String, nullable=False)
+
+    @staticmethod
+    def validate(kwargs):
+        schema.Schema({
+            'username': basestring,
+            'hostname': basestring,
+            'password': basestring,
+        }).validate(kwargs)
+
+    def session(self):
+        """ required from superclass. """
+        return self
+
+    def disconnect(self):
+        """ required from superclass. """
+
+    @staticmethod
+    def validate_port_name(port):
+        """ This driver accepts any string as port name."""
+
+    def str2list(self, a_string):
+        """ Converts a string representation of list to list.
+        Empty list is put as None.
+        """
+        if a_string[0] == '[' and a_string[1] == ']':
```

Comment:
if a_string is just `[]` then `return []` instead of calling `ast.literal_eva`l at all. And also, you can do the comparison only once (a_string[0] == '[') 

```python
if a_string[0] == '[':
        if a_string[1] == ']':
            return []
        a_string = a_string.replace("[", "['").replace("]", "']")
        a_string = a_string.replace(",", "','")
        a_list = ast.literal_eval(a_string)
        a_list = [ele.strip() for ele in a_list]
        return a_list
    else:
        return None
```

Reply:
re-read your docstring. if you want to return None for an empty list rather than just [], then this would be simpler. What do you think?
```python
if a_string is '[]':
    return None
# otherwise continue to parse it and then return it
return a_list

Reply:
@naved001, `is` isn't what you want, it checks pointer identity. You'd want `==` instead.

Reply:
oh cool, thanks for the correction!

Reply:
fixed the function. thanks for the keen observation.